src/main/java/org/elasticsearch/index/engine/internal/InternalEngine.java

```diff
@@ -743,7 +754,7 @@ public void flush(FlushType type, boolean force, boolean waitIfOngoing) throws E
                         { // commit and close the current writer - we write the current tanslog ID just in case
                             final long translogId = translog.currentId();
                             indexWriter.setCommitData(Collections.singletonMap(Translog.TRANSLOG_ID_KEY, Long.toString(translogId)));
-                            indexWriter.commit();
+                            commitIndexWriter(indexWriter);
```

Comment:
I think it's a problem to call fail engine here as we're holding the write lock. failEngine has a code block that acquires the read lock: 

```
} finally {
                closedOrFailed = true;
                try (InternalLock _ = readLock.acquire()) {
                    // we take the readlock here to ensure nobody replaces this IW concurrently.
                    if (indexWriter != null) {
                        indexWriter.rollback();
                    }
                } catch (Throwable t) {
                    logger.warn("Rolling back indexwriter on engine failure failed", t);
                    // to be on the safe side we just rollback the IW
                }
            }
```

Maybe we should reset indexWriter to null and then skip the read lock in failEngine if that's the case. Not sure.